Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/stitch_borders.h

```diff
@@ -572,57 +559,152 @@ std::size_t stitch_boundary_cycles(PolygonMesh& pm,
+      break;
 
-      if(stitched_hedges.count(h) > 0) // already treated
-        continue;
+    CGAL_assertion(is_border(start_h, pm));
 
-      std::vector<std::pair<halfedge_descriptor, halfedge_descriptor> > hedges_to_stitch;
+    // Associate as many consecutive halfedge pairs as possible ("zipping")
+    std::vector<std::pair<halfedge_descriptor, halfedge_descriptor> > hedges_to_stitch;
 
-      halfedge_descriptor hn = next(h, pm);
-      bool do_stitching = true;
-      do
+    halfedge_descriptor curr_h = start_h;
+    halfedge_descriptor curr_hn = next(curr_h, pm);
+    for(;;) // while we can expand the zipping range
+    {
+      // Don't create an invalid polygon mesh, even if the geometry allows it
+      if(face(opposite(curr_h, pm), pm) == face(opposite(curr_hn, pm), pm))
       {
-        hedges_to_stitch.push_back(std::make_pair(h, hn));
-        stitched_hedges.insert(h);
-        stitched_hedges.insert(hn);
+        unstitchable_halfedges.insert(curr_h);
+        bh = curr_hn;
+        break;
+      }
 
-        if(next(hn, pm) == h)
-          break;
+      CGAL_assertion(is_border(curr_h, pm));
+      CGAL_assertion(is_border(curr_hn, pm));
 
-        h = prev(h, pm);
-        hn = next(hn, pm);
+      hedges_to_stitch.push_back(std::make_pair(curr_h, curr_hn));
 
-        if(get(vpm, source(h, pm)) != get(vpm, target(hn, pm)))
-          do_stitching = false;
+      // check if we have reached the end of the cycle
+      if(prev(curr_h, pm) == curr_hn || prev(curr_h, pm) == next(curr_hn, pm))
+      {
+        bh = null_h;
+        break;
       }
-      while(do_stitching);
 
-      internal::stitch_borders_impl(pm, hedges_to_stitch);
-      ++stitched_boundary_cycles_n;
+      curr_h = prev(curr_h, pm);
+      curr_hn = next(curr_hn, pm);
+
+      // check if the next two halfedges are no longer geometrically compatible
+      if(get(vpm, source(curr_h, pm)) != get(vpm, target(curr_hn, pm)) ||
+         is_degenerate_edge(edge(curr_hn, pm), pm, parameters::vertex_point_map(vpm)))
+      {
+        bh = curr_hn;
+        break;
+      }
+    }
+
+    // bh must be a boundary halfedge on the border that will not be impacted by any stitching
+    typedef std::pair<halfedge_descriptor, halfedge_descriptor> Halfedge_pair;
+
+    CGAL_assertion_code(if(bh != null_h) {)
+    CGAL_assertion_code(  BOOST_FOREACH(const Halfedge_pair& hp, hedges_to_stitch) {)
+    CGAL_assertion(         bh != hp.first && bh != hp.second);
+    CGAL_assertion_code(}})
+
+    if(!hedges_to_stitch.empty())
+    {
+      std::size_t local_stitches = internal::stitch_borders_impl(pm, hedges_to_stitch);
+      stitched_boundary_cycles_n += local_stitches;
+
+      if(local_stitches == 0) // refused to stitch this halfedge pair range due to manifold issue
+      {
+        BOOST_FOREACH(const Halfedge_pair& hp, hedges_to_stitch)
+          unstitchable_halfedges.insert(hp.first);
```

Comment:
why only first?

Reply:
In the loop that tries to find a starting stitchable pair (above), the geometries of `hn` and `hnn` are compared and I need only to check that at least one of them is not marked as `unstitchable` (both is just redundant information). I chose the first one arbitrarily.

Actually, this probably could simply be `unstitchable_halfedges.insert(hedges_to_stitch.front().first);` because the other halfedges pairs are not formed by consecutive halfedges so these pairs will never be considered in the starting loop.